rust/src/daba/mod.rs

```diff
@@ -0,0 +1,151 @@
+use crate::FifoWindow;
+use alga::general::AbstractMonoid;
+use alga::general::Operator;
+use std::collections::VecDeque;
+use std::marker::PhantomData;
+
+#[derive(Clone)]
+pub struct DABA<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    // ith oldest value in FIFO order stored at vi = vals[i]
+    vals: VecDeque<Value>,
+    aggs: VecDeque<Value>,
```

Comment:
We implemented the `vals` and `aggs` using a single underlying queue. We created an internal struct that wrapped both a value and a partial aggregate, and then made the queue contain that struct. I think that will probably have better performance, both because of locality and by just doing less total work.

Reply:
Good point, that I should definitely fix

Reply:
I still do not think I fully get this algorithm, but thankfully it was pretty easy to implement 😅